yarn-project/simulator/src/public/public_tx_context.ts

```diff
@@ -0,0 +1,290 @@
+import {
+  type AvmProvingRequest,
+  type MerkleTreeReadOperations,
+  type PublicExecutionRequest,
+  type SimulationError,
+  type Tx,
+  TxExecutionPhase,
+} from '@aztec/circuit-types';
+import {
+  CombinedConstantData,
+  Fr,
+  Gas,
+  type GasSettings,
+  type GlobalVariables,
+  PublicAccumulatedDataArrayLengths,
+  type PublicCallRequest,
+  type PublicKernelCircuitPublicInputs,
+  PublicValidationRequestArrayLengths,
+  RevertCode,
+  type StateReference,
+} from '@aztec/circuits.js';
+import { type DebugLogger, createDebugLogger } from '@aztec/foundation/log';
+
+import { assert } from 'console';
+import { inspect } from 'util';
+
+import { AvmPersistableStateManager } from '../avm/index.js';
+import { DualSideEffectTrace } from './dual_side_effect_trace.js';
+import { PublicEnqueuedCallSideEffectTrace } from './enqueued_call_side_effect_trace.js';
+import { type WorldStateDB } from './public_db_sources.js';
+import { PublicSideEffectTrace } from './side_effect_trace.js';
+import { getCallRequestsByPhase, getExecutionRequestsByPhase, getPublicKernelCircuitPublicInputs } from './utils.js';
+
+class PhaseStateManager {
+  private currentlyActiveStateManager: AvmPersistableStateManager | undefined;
+
+  constructor(private readonly txStateManager: AvmPersistableStateManager) {}
+
+  fork() {
+    assert(!this.currentlyActiveStateManager, 'Cannot fork when already forked');
+    this.currentlyActiveStateManager = this.txStateManager.fork();
+  }
+
+  getActiveStateManager() {
+    return this.currentlyActiveStateManager || this.txStateManager;
+  }
+
+  isForked() {
+    return !!this.currentlyActiveStateManager;
+  }
+
+  mergeForkedState() {
+    assert(this.currentlyActiveStateManager, 'No forked state to merge');
+    this.txStateManager.mergeForkedState(this.currentlyActiveStateManager!);
+    // Drop the forked state manager now that it is merged
+    this.currentlyActiveStateManager = undefined;
+  }
+
+  discardForkedState() {
+    assert(this.currentlyActiveStateManager, 'No forked state to discard');
+    this.txStateManager.rejectForkedState(this.currentlyActiveStateManager!);
+    // Drop the forked state manager. We don't want it!
+    this.currentlyActiveStateManager = undefined;
+  }
+}
+
+export class PublicTxContext {
+  private log: DebugLogger;
+
+  private currentPhase: TxExecutionPhase = TxExecutionPhase.SETUP;
```

Comment:
As I mention later, ideally only one of PublicTxContext/PublicTxSimulator knows about the current phase and the phase flows. However I haven't given it a lot of thought to how difficult this might be.